#!/usr/bin/env Rscript
# Created by Michal Bukowski (michal.bukowski@tuta.io) under GPL-3.0 license

# This script is meant to function as a part of a Snakemake pipeline.
# R script for differential gene expression utilising Bioconductor DESeq2 and
# dedicated for quant files produced by Salmon for Illumina RNA-Seq squencing
# results for reversed stranded libraries.

# If you find it useful, please cite:
# Bukowski M et al. (2022) Staphylococcal saoABC operon codes for
# a DNA-binding protein SaoC implicated in the response to nutrient deficit.
# International Journal of Molecular Sciences, 23(12), 6443.

# Import libraries
suppressMessages(library("DESeq2"))
suppressMessages(library("tximport"))
suppressMessages(library("optparse"))

# Parse the arguments:
# --meta   -- a TSV file with metadata assigning quant files from one strain and
#             setting to different experimetnal groups that are to be compared
# --ref    -- group name that is meant to be used as reference in DGE
# --output -- a TSV file with DGE output data
option_list = list(
    make_option(c("--meta"), type="character", default=NULL,
                help="Metadata that describe samples (quant files)",
                metavar="character"),
    make_option(c("--ref"), type="character", default=NULL,
                help="Reference group name for comparisons",
                metavar="character"),
    make_option(c("--output"), type="character", default=NULL,
                help="Output differential gene expression table",
                metavar="character")
)

opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser)

# Load metadata generated by scripts/collect.py
metaData <- read.csv( opt$meta, header=TRUE, row.names=1, sep="\t")

# Give paths of Salmon quant files the names of replicas
files = metaData$fpath
names(files) <- rownames(metaData)

# Import Salmon quant files
txi <- tximport(files, type="salmon", txOut=TRUE)

# Perform DGE among groups using one as the reference group
dds <- DESeqDataSetFromTximport(txi, metaData["group"], ~group)
dds <- dds[ rowSums( counts(dds) ) >= 10, ]
dds$group <- relevel(dds$group, ref=opt$ref)
dds <- DESeq(dds)

# Process the results assuming FDR of 0.05 and perfomr shirinking. Format
# the results and save them to a TSV file
res <- results(dds, alpha=0.05)
res <- lfcShrink(dds, res=res, coef=2, type="apeglm")
summary(res)
skipped <- setdiff(rownames(txi$counts), rownames(res))
res[skipped,] <- NA
res <- res[ order(row.names(res)), ]
res <- cbind(rownames(res), data.frame(res, row.names=NULL))
names(res)[1] <- "locus_tag"
write.table(res, file=opt$output, sep="\t", row.names=FALSE, quote=FALSE)


